### PR TITLE
#217: support AUTHORIZATION and CHARGE for all payment types

### PR DESCRIPTION
**Pull request by andrii-kovalenko-ct**
as a hot fix for https://github.com/commercetools/commercetools-payone-integration/issues/217:

Now we don't validate transaction type: the customer should create proper transaction type on theirs own, the service simply executes transaction on payone, if it is possible.

- [`PaymentMethods`](https://github.com/commercetools/commercetools-payone-integration/pull/218/files#diff-3439e6d93c20fc6f9d51971b99d0cd1e)
  - removed `AUTHORIZATION`/`CHARGE` limits for every payment method
  - define global `supportedTransactionTypes` set of supported types

- [`BankTransferAuthorizationRequest/BankTransferRequest `](https://github.com/commercetools/commercetools-payone-integration/pull/218/files#diff-c0d8899cbde8d8d068967980a34c0a3c): make it flexible, accepting request type as a constructor arguments.
Same for [`BankTransferInAdvancePreautorizationRequest/BankTransferInAdvanceRequest`](https://github.com/commercetools/commercetools-payone-integration/pull/218/files#diff-b7b6c7037201d91d340ee35ee1a0e8d5)

- [`SofortBankTransferRequestFactory `](https://github.com/commercetools/commercetools-payone-integration/pull/218/files#diff-cc21a529ce437cdf3420cbcdf4e7322e): using default basic `super.createPreauthorizationRequest/super.createAuthorizationRequest` just patch them additionally with IBAN/BIC fields where it is necessary.

- [`TenantFactory`](https://github.com/commercetools/commercetools-payone-integration/pull/218/files#diff-d34fd6b71077118ff85d598f7827ba76): 
  - because of bad design from the beginning and instantiation values inside constructor (instead of DI), we have to re-factor this class slowly from PR to PR. This time we make `PayonePostService` mocking/injection possible. See also [`tenantFactoryTest`](https://github.com/commercetools/commercetools-payone-integration/pull/218/files#diff-933c751c42680c69e2d881da80d71627R242) for more details.
  - [remove duplication+hardcoding of supported payment methods](https://github.com/commercetools/commercetools-payone-integration/pull/218/files#diff-d34fd6b71077118ff85d598f7827ba76L208), use `supportedPaymentMethods` and `supportedTransactionTypes` static sets instead

- added [Unit test](https://github.com/commercetools/commercetools-payone-integration/pull/218/files#diff-933c751c42680c69e2d881da80d71627R168) which verifies all the payment methods could process all the expected CTP transaction types (`AUTHORIZATION` and `CHARGE`)

- Some `@Nonnull` annotation added, classes renamed to reflect better current workflow

### CODE REVIEW

**Comment by andrii-kovalenko-ct** [comment]
@heshamMassoud 
not really, i'd say it was never expected to be validated at all, so basically we do opposite job: we fix the service so it works respectively to current documentation

File: service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java [hunk 1]
```diff
@@ -22,11 +21,6 @@ public PaymentDispatcher(final Map<PaymentMethod, PaymentMethodDispatcher> metho
         this.payoneInterfaceName = payoneInterfaceName;
     }
 
-    @Override
-    public void accept(PaymentWithCartLike paymentWithCartLike) {
-        dispatchPayment(paymentWithCartLike);
-    }
```
Comment (andrii-kovalenko-ct) [inline service/src/main/java/com/commercetools/pspadapter/payone/PaymentDispatcher.java]: It just was never used

File: service/src/test/java/com/commercetools/pspadapter/tenant/TenantFactoryTest.java [hunk 2]
```diff
@@ -131,14 +154,121 @@ public void createPaymentDispatcher() throws Exception {
...
+        };
+    }
+
+    /**
+     * <table border="1">
+     * <tr><th>CTP</th><td>&nbsp;</td><th>Payone</th></tr>
+     * <tr><td>{@link TransactionType#AUTHORIZATION}</td><td>&nbsp;-&gt;&nbsp;</td><td>{@link RequestType#PREAUTHORIZATION}</td></tr>
+     * <tr><td>{@link TransactionType#CHARGE}</td><td>&nbsp;-&gt;&nbsp;</td><td>{@link RequestType#AUTHORIZATION}</td></tr>
+     * <tr><td>other</td><td>&nbsp;-&gt;&nbsp;</td><td>"unexpected transaction request type "%s""</td></tr>
+     * </table>
```
Comment (LEQADA) [inline service/src/test/java/com/commercetools/pspadapter/tenant/TenantFactoryTest.java]: Maybe let's add some human-readable comments too?

Comment (andrii-kovalenko-ct) [reply service/src/test/java/com/commercetools/pspadapter/tenant/TenantFactoryTest.java]: Not sure what you mean. Javadoc actually are human specific comments:
image

File: service/src/main/java/com/commercetools/pspadapter/payone/mapping/BanktTransferInAdvanceRequestFactory.java [hunk 3]
```diff
@@ -24,16 +27,27 @@ public BanktTransferInAdvanceRequestFactory(@Nonnull final TenantConfig tenantCo
     }
 
     @Override
-    public BankTransferInAdvancePreautorizationRequest createPreauthorizationRequest(PaymentWithCartLike paymentWithCartLike) {
-        final Payment ctPayment = paymentWithCartLike.getPayment();
+    @Nonnull
```
[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply service/src/main/java/com/commercetools/pspadapter/payone/mapping/BanktTransferInAdvanceRequestFactory.java]: fb0ac6ae

File: service/src/main/java/com/commercetools/pspadapter/payone/domain/ctp/paymentmethods/PaymentMethod.java [hunk 4]
```diff
@@ -4,98 +4,94 @@
...
+
+    /**
+     * Set of supported CTP transaction types.
+     * <p>
+     * <b>Note:</b> solving issue
+     * <i><a href="https://github.com/commercetools/commercetools-payone-integration/issues/217">
+     * Make transaction type handling transparent</a></i>
+     * we support any {@link TransactionType#AUTHORIZATION AUTHORIZATION} or {@link TransactionType#CHARGE CHARGE}
+     * transaction for all payment methods: the Payone request will be sent without additional validation.
+     * It is a responsibility of
```
[the remark this answers is not in the document]
Comment (andrii-kovalenko-ct) [reply service/src/main/java/com/commercetools/pspadapter/payone/domain/ctp/paymentmethods/PaymentMethod.java]: f55768e2

**Comment by andrii-kovalenko-ct** [comment]
also resolves #126 (#131)

**Pull request merged**